### PR TITLE
pkgs.formats.toml: fix TOML semantics by upgrading tomlkit

### LINKED ISSUE #237521: pkgs.formats.toml changed semantics in 23.05
### Describe the bug
`pkgs.formats.toml` changed behaviour between nixos 22.11 and NixOS 23.05, apparently producing semantically different outputs (which broke our telegraf module)

```
~/nixpkgs nixos-22.11
❯ git rev-parse HEAD
eef86b8a942913a828b9ef13722835f359deef29

~/nixpkgs nixos-22.11
❯ cat $(nix-build --expr '(((import ./. {}).formats.toml {}).generate "test" { processors = { override = [ { tags = { cluster = "staging"; }; } ]; rename = [ { replace = [ { dest = "ceph_telegraf_check"; measurement = "exec"; } ]; }]; }; })')
[processors]
[[processors.override]]
[processors.override.tags]
cluster = "staging"

[[processors.rename]]
[[processors.rename.replace]]
dest = "ceph_telegraf_check"
measurement = "exec"

~/nixpkgs nixos-22.11
❯ git checkout nixos-23.05

~/nixpkgs nixos-23.05
❯ git rev-parse HEAD
d3bb401dcfc5a46ce51cdfb5762e70cc75d082d2

~/nixpkgs nixos-23.05
❯ cat $(nix-build --expr '(((import ./. {}).formats.toml {}).generate "test" { processors = { override = [ { tags = { cluster = "staging"; }; } ]; rename = [ { replace = [ { dest = "ceph_telegraf_check"; measurement = "exec"; } ]; }]; }; })')

[processors]
[processors.override.tags]
cluster = "staging"
[[processors.rename.replace]]
dest = "ceph_telegraf_check"
measurement = "exec"
```

Bisect implies this is the commit that broke(?) remarshal: 620c7eaeeeaed47156d51b04e50ea220614cfa62

Was this intentional?

### PR DESCRIPTION
**Pull request by luizribeiro**
###### Description of changes

This fixes #237521 by upgrading tomlkit to 0.11.8.

It also introduces tests to make sure this behavior is kept, as some of our configurations (such as telegraf) depend on it.

Tested these changes by running `nix build .\#tests.pkgs-lib.formats`.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/tomlkit/default.nix
Comment (infinisil) [inline pkgs/development/python-modules/tomlkit/default.nix:13]: Downgrading packages should be exceptional: Is the new behavior a bug or a feature?
- If it's a feature we should either
  - Introduce a new `tomlkit_11_4` attribute and use that, or
  - Define the older version internally to the `pkgs.formats` code (this way we don't have a new attribute to maintain)  
- If it's a bug we should add a comment here saying that we're waiting for the bug to get fixed before upgrading.

**Review by infinisil — CHANGES_REQUESTED** [review]
Also, does this need to be backported?

File: pkgs/development/python-modules/tomlkit/default.nix
Comment (luizribeiro) [reply pkgs/development/python-modules/tomlkit/default.nix:13]: You're absolutely right :)

It turns out that tomlkit 0.11.8 (the latest version) doesn't have this bug either. So doing an upgrade and introducing the tests seems way more appropriate.

The challenge is that tomlkit is now built with poetry and.... poetry also depends on tomlkit.

So I think until we bootstrap poetry itself, we need to keep 0.11.6 as `tomlkit_11_6`, similar to your suggestion.

I'll make those changes. Hopefully I don't mess it up :-)

Thanks for the quick review, btw!

**Comment by luizribeiro** [comment]
Not sure if this is the cleanest way to go about this, so I'm open to suggestions.

> Also, does this need to be backported?

We need to have it merged into the 23.05 release, if that's what you are asking (sorry, I'm not yet very familiar with the nixpkgs release process / terminology).

File: pkgs/development/python-modules/poetry-core/default.nix
Comment (infinisil) [inline pkgs/development/python-modules/poetry-core/default.nix:59]: That's not right, there's nothing related to tomlkit newly depending on poetry in the 0.11.7 release: https://github.com/sdispater/tomlkit/compare/0.11.6...0.11.7

The error you ran into must've been this one:
```
this derivation will be built:
  /nix/store/cgyhiwnccdb2lh4q3di996iv35yk4hdf-python3.10-tomlkit-0.11.7.drv
building '/nix/store/cgyhiwnccdb2lh4q3di996iv35yk4hdf-python3.10-tomlkit-0.11.7.drv'...
[...]
building
Executing setuptoolsBuildPhase
Traceback (most recent call last):
  File "/build/tomlkit-0.11.7/nix_run_setup", line 8, in <module>
    exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\\r\\n', '\\n'), __file__, 'exec'))
  File "/nix/store/jhflvwr40xbb0xr6jx4311icp9cym1fp-python3-3.10.12/lib/python3.10/tokenize.py", line 394, in open
    buffer = _builtin_open(filename, 'rb')
FileNotFoundError: [Errno 2] No such file or directory: 'setup.py'
/nix/store/7aprjn0yf6p06krg4087sxvilfp4c8yy-stdenv-linux/setup: line 1596: pop_var_context: head of shell_variables not a function context
```

For some reason there was a `setup.py` before but now there isn't anymore. I could reproduce this by just fetching the `tomlkit-0.11.6.tar.gz` from https://pypi.org/project/tomlkit/0.11.6/#files, and comparing it to the one from https://pypi.org/project/tomlkit/0.11.7/#files. The latter indeed doesn't have a `setup.py` file anymore.

I really don't know why this changed or why the `setup.py` was even generated originally, but this looks like some upstream inconsistency which shouldn't have happened.

Can you file an upstream bug for that? I think we should just downgrade to 0.11.6 until that's resolved then.

Comment (luizribeiro) [reply pkgs/development/python-modules/poetry-core/default.nix:59]: > but this looks like some upstream inconsistency which shouldn't have happened.

Why do you say so? They're using `pyproject.toml` to manage their packaging, so there's no need for a `setup.py`. In fact, the PyPA's sample project only has `pyproject.toml` as of last year: https://github.com/pypa/sampleproject/commit/aeeb50a948addcb712ad4261df472263514991e1

They might not have announced it as a breaking change on 0.11.7 so, sure, it might be an inconsistency. But I don't think it's worth a bug report upstream unless I'm missing something.

Comment (luizribeiro) [reply pkgs/development/python-modules/poetry-core/default.nix:59]: > The error you ran into must've been this one:

Not quite, the error you are referring to is the error you will get if you build `tomlkit` with `format` as `setuptools` instead of `pyproject` (which was also updated on this PR)

Comment (infinisil) [reply pkgs/development/python-modules/poetry-core/default.nix:59]: Something changed so that no `setup.py` is generated anymore, and this change is not recorded in the git history nor documented anywhere. We don't really know what occurred there, and we shouldn't try to work around it with more code unless we know it's the right solution. Maybe `setup.py` was explicitly provided to make bootstrapping poetry easier?

Comment (luizribeiro) [reply pkgs/development/python-modules/poetry-core/default.nix:59]: Filed https://github.com/sdispater/tomlkit/issues/305

Will revert this PR to the original PR with a comment linking to that issue.

Comment (infinisil) [reply pkgs/development/python-modules/poetry-core/default.nix]: Let's wait a bit to see if the issue gets resolved (I see you found a good lead)

Comment (luizribeiro) [reply pkgs/development/python-modules/poetry-core/default.nix]: @infinisil, would you be open to us landing this as is? Telegraf is currently pretty much unusable because of the `pkgs.format.toml` issue.

I'm also not super sure that setting `tool.poetry.build.generate-setup-file` upstream is a reasonable solution, since that option is going to be removed from `poetry` in the future (see https://python-poetry.org/blog/announcing-poetry-1.3.0/#generate-setup-file--false).

A solution might be for us to allow for Python packages with the `pyproject` `format` to be built with `pip` instead of `poetry` within `nixpkgs`. But I haven't looked into how that would happen yet.

Comment (luizribeiro) [reply pkgs/development/python-modules/poetry-core/default.nix]: We have these options now:

1. Pin to 0.11.4 (current state of this PR)
2. Upgrade to 0.11.8 and apply a patch which starts generating `setup.py`
3. Upgrade to 0.11.8 and apply a patch which moves from using `poetry` to using `setuptools` (see my most recent comment on https://github.com/sdispater/tomlkit/pull/306)
4. Wait until a new release of `tomlkit` comes out with a fix

I'm personally leaning towards 2, since it's the least intrusive change to `nixpkgs` (and hence simpler to remove once we figure out the solution upstream).

What do you think?

File: pkgs/development/python-modules/tomlkit/default.nix
Comment (infinisil) [inline pkgs/development/python-modules/tomlkit/default.nix:16]: We downgrade to <0.11.6 because of the changed semantics in versions 0.11.6 to [are the semantics fixed in 0.11.7 or 0.11.8?], and we don't upgrade to 0.11.8 because of the `setup.py` issue.

We could also try just downgrade to 0.11.5, there's nothing wrong with that version afaik.

**Review by infinisil — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/poetry-core/default.nix
Comment (infinisil) [reply pkgs/development/python-modules/poetry-core/default.nix]: Oh yeah good point, I like number 2, but still I'd like to give upstream some time to respond. If you need the fix immediately you can still apply it as a patch to Nixpkgs, or use overlays.

File: pkgs/development/python-modules/tomlkit/default.nix
Comment (luizribeiro) [reply pkgs/development/python-modules/tomlkit/default.nix:16]: I feel like upgrading to 0.11.8 with a patch would be best. Are you opposed to that?

**Comment by luizribeiro** [comment]
Alright I'm calling it a day on this for now. Thanks for the help and sorry for all the churn :)

I've added a commit to my branch which has option 3 implemented that I mentioned above (a patch which makes `tomlkit` be built with `setuptools` instead of `poetry`).

I'll squash / drop those commits depending on the solution we go with before we merge this in.

**Comment by luizribeiro** [comment]
After discussing this upstream on https://github.com/sdispater/tomlkit/issues/305, it turns out that `poetry-core` should not depend on `tomlkit`. `poetry-core` has no dependencies on other libraries and should be used to build `tomlkit` instead of `poetry`.

I've updated this PR accordingly.

**Review by infinisil — CHANGES_REQUESTED** [review]
Nice! Could you split this into separate commits like this?
1. Remove poetry dependency on tomlkit and set `format = "pyproject" on tomlkit
2. Update tomlkit to 0.11.8
3. Introduce the test case

**Review by infinisil — APPROVED** [review]
Awesome thanks!

**Comment by mweinelt** [comment]
f68995642f6efc4fb0d4b208a5e90864237407ba broke pylint and by extension all nixos tests.

**Comment by vcunat** [comment]
Note that this PR was tagged by OfBorg as very large rebuild and thus it should not have been merged to master.

**Comment by infinisil** [comment]
Ah sorry I forgot to pay attention the label. Here's [a Hydra failure](https://hydra.nixos.org/build/229359940) for reference, also fails on `x86_64-linux`. After some quick error searching it looks like [this commit](https://github.com/sdispater/tomlkit/commit/9cf42c61d473fc3fd63b1b5827f4335fb0748f71) in tomlkit introduced that error, and it's only included in 0.11.8, not 0.11.7. I can confirm that only upgrading to 0.11.7 fixes the pylint build.

**Comment by mweinelt** [comment]
Picked into #244135.

**Pull request merged**
